src/devices/Mcp3428/Mcp342x.cs

```diff
@@ -0,0 +1,387 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Device.I2c;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Iot.Device.Mcp3428
+{
+    public abstract class Mcp342x : IDisposable
+    {
+        /// <summary>
+        /// Protected constructor to easily generate MCP3426/7 devices whose only difference is channel count and I2C address
+        /// </summary>
+        /// <param name="i2CDevice">The i2 c device.</param>
+        /// <param name="channels">The channels.</param>
+        /// <autogeneratedoc />
+        protected Mcp342x(I2cDevice i2CDevice, int channels)
+        {
+            _i2cDevice = i2CDevice;
+            ChannelCount = channels;
+            ReadValue(); // Don't like this in constructor, makes sure props are valid
+        }
+
+        /// <summary>
+        /// Gets the last transmitted bytes. Debug function
+        /// </summary>
+        /// <value>The last bytes.</value>
+        /// <autogeneratedoc />
+        public ReadOnlyMemory<byte> LastBytes => _readBuffer;
+
+        /// <summary>
+        /// Channel most recently read
+        /// </summary>
+        /// <value>The last channel.</value>
+        /// <autogeneratedoc />
+        public byte LastChannel => LastConversion.Channel;
+
+        public ModeEnum Mode
+        {
+            get => _mode;
+            set
+            {
+                WriteConfig(Helpers.SetModeBit(LastConfigByte, value));
+                _mode = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the input gain.
+        /// </summary>
+        /// <value>The pga gain.</value>
+        /// <autogeneratedoc />
+        public GainEnum PGAGain
+        {
+            get => _pgaGain;
+            set
+            {
+                WriteConfig(Helpers.SetGainBits(LastConfigByte, value));
+                _pgaGain = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the bit resolution of the result.
+        /// </summary>
+        /// <value>The resolution.</value>
+        /// <autogeneratedoc />
+        public ResolutionEnum Resolution
+        {
+            get => _resolution;
+            set
+            {
+                WriteConfig(Helpers.SetResolutionBits(LastConfigByte, value));
+                _resolution = value;
+            }
+        }
+
+        public void Dispose()
```

Comment:
we usually implement this as:

```
_i2cDevice?.Dispose();
_i2cDevice = null;
```

Reply:
Was thinking if I got it from outside then I shouldn't dispose it.

Reply:
let's do it for consistency - we will likely re-iterate on the dispose pattern in the future and fix it everywhere if needed

Reply:
I still don't like it but it's there now. 
Should I remove readonly modifier from the field to be able to set it to null?

Reply:
let's leave it as is